packages/protocol/src/utils/event.ts

```diff
@@ -0,0 +1,92 @@
+/********************************************************************************
+ * Copyright (c) 2023 EclipseSource and others.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License v. 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0.
+ *
+ * This Source Code may also be made available under the following Secondary
+ * Licenses when the conditions for such availability set forth in the Eclipse
+ * Public License v. 2.0 are satisfied: GNU General Public License, version 2
+ * with the GNU Classpath Exception which is available at
+ * https://www.gnu.org/software/classpath/license.html.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
+ ********************************************************************************/
+import * as jsonrpc from 'vscode-jsonrpc';
+import { Disposable } from './disposable';
+import { SafeFunction } from './type-util';
+
+/**
+ * GLSP event/emitter API. Reuses the API provided by
+ * `vscode-jsonrpc` and overrides certain type definitions to match the type
+ * definitions provided by GLSP (Event, Disposable)
+ */
+
+/**
+ * Represents a typed event. Listener functions can subscribe for a event
+ * and will be called when the event is fired. Typically events are provided by services via
+ * public property. For example assume we have a service `myService` that exposes a
+ * `onStartupComplete` event. Then a listener could be registered like this
+ * ```ts
+ *      const disposable= this.myService.onStartupComplete(event=> doSomething(event));
+ *
+ *     // If the listener should be unregistered again, simply call
+ *      disposable.dispose()
+ * ```
+ */
+export interface Event<T> extends jsonrpc.Event<T> {
+    /**
+     *
+     * @param listener The listener function will be called when the event happens.
+     * @param thisArgs The 'this' which will be used when calling the event listener.
+     * @param disposables An array to which the {@link Disposable} for removing the listener will be added.
+     * @returns a {@link Disposable} to remove the listener again.
+     */
+    (listener: (e: T) => any, thisArgs?: any, disposables?: Disposable[]): Disposable;
```

Comment:
So I assume we override this because we ant to use our own "Disposable", right? I'm just wondering because at the moment our own disposable does not have any extra functionality. Also, should our Disposable now also extend `jsonrpc.Disposable`?

I think I'm just a bit confused as to why we need our own types that are basically the same (with the exception of exchanging `Function` with `SafeFunction` in the EmitterOptions.

Reply:
We don't strictly need our own types this has been added to be more future proof and for developer convenience.
By overriding with our own types we ensure that things like auto import etc. use the correct types as well (from @eclipse-glsp/client and not vscode-jsonrpc).
In addition, it gives us the flexibility to add custom functionality to our types later on and also ensure that we could move away from vscode-jsonrpc if we decide at some point that we don't want to use it anymore.
Another indirect benefit is that we can add more documentation to the event classes 😉 

Ensuring compatibility with the vscode disposable by extending sounds like a good idea to me.

Reply:
Thank you for the clarification! I expected as much but I just wanted to make sure I fully understand the intention behind it. With that, this looks totally fine to me.